src/internal/pjs/api_server.go

```diff
@@ -0,0 +1,110 @@
+package pjs
+
+import (
+	"context"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/dbutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pjsdb"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset"
+	txnenv "github.com/pachyderm/pachyderm/v2/src/internal/transactionenv"
+	"github.com/pachyderm/pachyderm/v2/src/pjs"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type apiServer struct {
+	pjs.UnimplementedAPIServer
+	env    Env
+	txnEnv *txnenv.TransactionEnv
+}
+
+func newAPIServer(env Env) *apiServer {
+	return &apiServer{
+		env:    env,
+		txnEnv: env.TxnEnv,
+	}
+}
+
+func (a *apiServer) CreateJob(ctx context.Context, request *pjs.CreateJobRequest) (response *pjs.CreateJobResponse, retErr error) {
+	var ret pjs.CreateJobResponse
+	if request.Input == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "missing data input")
+	}
+	if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+		program, err := fileset.ParseID(request.Program)
+		if err != nil {
+			return errors.EnsureStack(err)
+		}
+		var inputs []fileset.PinnedFileset
+		for _, input := range request.Input {
+			inputID, err := fileset.ParseID(input)
+			if err != nil {
+				return errors.EnsureStack(err)
+			}
+			inputs = append(inputs, fileset.PinnedFileset(*inputID))
+		}
+		req := pjsdb.CreateJobRequest{
+			Program: fileset.PinnedFileset(*program),
+			Inputs:  inputs,
+		}
+		jobID, err := pjsdb.CreateJob(ctx, sqlTx, req)
+		if err != nil {
+			return errors.EnsureStack(err)
+		}
+		ret.Id = &pjs.Job{Id: int64(jobID)}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return &ret, nil
+}
+
+func (a *apiServer) InspectJob(ctx context.Context, req *pjs.InspectJobRequest) (*pjs.InspectJobResponse, error) {
+	var jobInfo pjs.JobInfo
+	if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+		job, err := pjsdb.GetJob(ctx, sqlTx, pjsdb.JobID(req.Job.Id))
+		if err != nil {
+			return errors.EnsureStack(err)
+		}
+		jobInfo = pjs.JobInfo{
+			Job: req.Job,
+			ParentJob: &pjs.Job{
+				Id: int64(job.Parent),
+			},
+			Program: job.Program.HexString(),
+			State:   pjs.JobState_QUEUED,
+		}
+		for _, filesetID := range job.Inputs {
+			jobInfo.Input = append(jobInfo.Input, filesetID.HexString())
+		}
+		if !job.Processing.IsZero() {
```

Comment:
82-88 here can be in a switch block.   In Go it's more idiomatic  to write sequences of mutually exclusive if-elses as a switch.

```go
switch {
case ifFirstCaseIsTrue:
case ifSecondCaseIsTrue:
default:
  return fmt.Errorf("invalid job from db: %v")
```

Reply:
@brendoncarroll 
line 88 handles job result if state is done. Line 82-87 is not mutually exclusive. A job in "Done" state also has processing timestamp and queued timestamp, so line 85 and 82 can be correct at the same time.

Reply:
A job can only be in one state at a time though.  `jobInfo.State` can only be assigned to one value after all is said and done.  If we determine what that is by checking timestamps, and there are some database states that are unaccounted for, then we should be able to detect when they occur.

Reply:
Got it. I'll fix this. I shouldn't assume all the states are valid.